keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problem whereas changes to
+available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
+
+When a new Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod. When a Pod resize is requested, Kubelet attempts
+to update the Pod and its Containers. If required, Kubelet may evict lower
+priority Pods from the Node to make room for the resize. Scheduler observes the
+resize request, and updates the accounting in its cache.
+
+The following steps denote a typical flow of an in-place resize operation for a Pod
+with ResizePolicy set to NoRestart for all its Containers.
+
+1. Initiating actor updates Pod's Container.ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements (e.g. limits are not below
+   requested resources, QoS class does not change, ResourceQuota not exceeded..).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers reject the update, API Server responds with an
+     appropriate error message.
+1. API Server updates PodSpec object with the new desired ResourceRequirements.
+1. Kubelet observes that Pod's Container.ResourceRequirements and
+   ContainerStatus.ResourcesAllocated differ. It checks its Node allocatable
+   resources to determine if the new desired ResourceRequirements fit the Node.
+   * _Case 1_: Kubelet finds new ResourceRequirements fit. It sets
+     ResizingPod.Status to true, and ResizingPod.Reason to InProgress, then
+     applies resized cgroup limits to the Pod and its Containers, and once
+     successfully done, updates Pod's ContainerStatus.ResourcesAllocated to
+     reflect the desired ResourceRequirements. It then sets ResizingPod.Status
+     to false, and ResizingPod.Reason to Success.
+     - If at the same time, a new Pod was scheduled against the capacity used by
+       this resource resize, that Pod is rejected during Kubelet admission if
+       Node has no more room.
+   * _Case 2_: Kubelet finds that desired ResourceRequirements does not fit.
+     - Kubelet checks to see if evicting lower priority Pods on the Node can
+       successfully resize the Pod. If yes, it sets ResizingPod.Status to true,
+       ResizingPod.Reason to PreEmpting, and initiates pre-emption of lower
```

Comment:
i dont think we need to advertise on a pod A that the kubelet may be pre-empting another pod B in order to satisfy a request.  i think its fine to just keep this simple and say that the kubelet is in-progress trying to achieve this.  its probably more important to state how we would capture that pod B was preeempted.  i assume it will be similar to any other eviction outcome but we would allow for graceful termination periods?

Reply:
Yes. We use the eviction subresource that @liggitt [suggested above](https://github.com/kubernetes/enhancements/pull/686#issuecomment-502145856). I believe this would safely evict the lower priority Pod while respecting PDBs.
I've captured this in the Notes section of the KEP.

@kgolab: In today's sig-node meeting, we discussed the idea of removing PreEmpting condition and collapsing it into InProgress. Does VPA lose any information if we did this change? 

There was also a suggestion that we may not need InProgress state - this can be implied if Container.Resources and ContainerStatus.ResourcesAllocated are different, and Kubelet didn't set it to FailedNodeCapacity. I'll walk through the flow to see if we can remove that as well, but please let me know if you have any reasons we should keep these.

Reply:
Eviction is not part of this KEP now. Resolving the issue.